core/src/main/scala/org/apache/spark/sql/delta/schema/SchemaUtils.scala

```diff
@@ -538,57 +538,53 @@ object SchemaUtils extends DeltaLogging {
-            prettyFieldName(stack ++ Seq(thisCol)))
-        case (Seq(), MapType(_, _, _)) =>
-          (Nil, 2)
-        case (Seq("key", _ @ _*), MapType(keyType: StructType, _, _)) =>
-          val (child, size) = find(colTail.tail, keyType, stack :+ thisCol)
-          (MAP_KEY_INDEX +: child, size)
-        case (Seq("key"), MapType(_, _, _)) =>
-          (Seq(MAP_KEY_INDEX), 0)
-        case (Seq("value", _ @ _*), MapType(_, valueType: StructType, _)) =>
-          val (child, size) = find(colTail.tail, valueType, stack :+ thisCol)
-          (MAP_VALUE_INDEX +: child, size)
-        case (Seq("value"), MapType(_, _, _)) =>
-          (Seq(MAP_VALUE_INDEX), 0)
-        case (_, MapType(_, _, _)) =>
-          throw DeltaErrors.foundMapTypeColumnException(
-            prettyFieldName(stack ++ Seq(thisCol, "key")),
-            prettyFieldName(stack ++ Seq(thisCol, "value")))
-        case (_, o) =>
-          if (column.length > 1) {
-            throw DeltaErrors.columnPathNotNested(columnPath, o, column)
+      resolver: Resolver = DELTA_COL_RESOLVER): Seq[Int] = {
+    def find(
+        searchPath: Seq[String],
+        currentType: DataType,
+        currentPath: Seq[String] = Nil): Seq[Int] = {
+      if (searchPath.isEmpty) return Nil
+
+      val currentFieldName = searchPath.head
+      val path = currentPath :+ currentFieldName
+      (currentType, currentFieldName) match {
+        case (struct: StructType, _) =>
+          lazy val columnPath = UnresolvedAttribute(path).name
+          val pos = struct.indexWhere(f => resolver(f.name, currentFieldName))
+          if (pos == -1) {
+            throw new IndexOutOfBoundsException(columnPath)
           }
-          (Nil, 0)
+          val childPath = find(searchPath.tail, struct(pos).dataType, path)
+          pos +: childPath
+
+        case (map: MapType, "key") =>
+          val childPath = find(searchPath.tail, map.keyType, path)
+          MAP_KEY_INDEX +: childPath
+
+        case (map: MapType, "value") =>
+          val childPath = find(searchPath.tail, map.valueType, path)
+          MAP_VALUE_INDEX +: childPath
+
+        case (_: MapType, _) =>
+          throw DeltaErrors.foundMapTypeColumnException(
+            prettyFieldName(currentPath :+ "key"),
+            prettyFieldName(currentPath :+ "value"))
+
+        case (array: ArrayType, "element") =>
+          val childPath = find(searchPath.tail, array.elementType, path)
+          ARRAY_ELEMENT_INDEX +: childPath
+
+        case (_: ArrayType, _) =>
+          throw DeltaErrors.incorrectArrayAccessByName(
+            prettyFieldName(currentPath :+ "element"),
+            prettyFieldName(currentPath))
+        case _ =>
+          throw DeltaErrors.columnPathNotNested(currentFieldName, currentType, currentPath)
       }
-      (Seq(pos) ++ children, lastSize)
     }
 
     try {
-      find(column, schema, Nil)
+      find(column, schema)
     } catch {
       case i: IndexOutOfBoundsException =>
```

Comment:
BTW, I'm not sure why we don't throw the better error inside find immediately. There's only one throw site. And I spent time typing up a comment about that this would be a terrible error, which I now had to delete because I found the better error. If you don't move this throw inside find, then at least add a comment there to say that it gets translated to a better error in the caller, to save the next person from triggering on this...